gadget/install/install.go

```diff
@@ -377,6 +376,140 @@ func Run(model gadget.Model, gadgetRoot, kernelRoot, bootDevice string, options
 	}, nil
 }
 
+// structureFromPartDevice returns the OnDiskStructure for a partition
+// node.
+func structureFromPartDevice(diskVol *gadget.OnDiskVolume, partNode string) (*gadget.OnDiskStructure, error) {
+	for _, p := range diskVol.Structure {
+		// We remove /dev from p.Node here
+		if p.Node == partNode {
+			return &p, nil
+		}
+	}
+
+	return nil, fmt.Errorf("cannot find partition %q", partNode)
+}
+
+// laidOutStructureForDiskStructure searches for the laid out structure that
+// matches a given OnDiskStructure.
+func laidOutStructureForDiskStructure(laidVols map[string]*gadget.LaidOutVolume, gadgetVolName string, onDiskStruct *gadget.OnDiskStructure) (
+	*gadget.LaidOutStructure, error) {
+
+	for _, laidVol := range laidVols {
+		// Check that this is the right volume
+		if laidVol.Name != gadgetVolName {
+			continue
+		}
+		for _, laidStruct := range laidVol.LaidOutStructure {
+			if onDiskStruct.Name == laidStruct.Name {
+				return &laidStruct, nil
+			}
+		}
+	}
+
+	return nil, fmt.Errorf("cannot find laid out structure for %q", onDiskStruct.Name)
+}
+
+// sysfsPathForBlockDevice returns the sysfs path for a block device.
+func sysfsPathForBlockDevice(device string) (string, error) {
+	syfsLink := filepath.Join("/sys/class/block", filepath.Base(device))
+	partPath, err := os.Readlink(syfsLink)
+	if err != nil {
+		return "", fmt.Errorf("cannot read link %q: %v", syfsLink, err)
+	}
+	// Remove initial ../../ from partPath, and make path absolute
+	return filepath.Join("/sys/class/block", partPath), nil
```

Comment:
Should we also call filepath.Clean() here? I guess it does not really matter but filepath.EvalSymlinks() also does that (but we can't use that AIUI because it's a bit silly and works relative to CWD).

Reply:
It should not be necessary because of the way we obtain the path - if it actually was it would be probably a bug